### DISCUSSION TITLE
Document required OIDC claims

### DISCUSSION THREAD

**Discussion by ghost** [outline/outline#3033] [Self Hosting]
Outline implicitly needs a `name` claim on the OIDC token.

https://github.com/outline/outline/blob/b52b1b02fe62b5e51a89935c1937d83724f9f45a/server/routes/auth/providers/oidc.ts#L102

While this is technically part of the [Standard Claims](https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims), many providers (like Keycloak) do not include it by default. There is also no error handling if this value is undefined, beyond a non-null constraint in the database.

A relatively easy fix would be to make them configurable with environment variables the same way `OIDC_USERNAME_CLAIM` already is and place the default in `.env.sample`.

**Comment by tommoor** [comment]
I think the answer is to change the KeyCloak config here, it doesn't make sense to have a "default" name or all your users are going to have the same name!

**Reply by ghost** [reply]
Ah, apologies for the confusion, I meant a default claim to pull out of the JWT. But simply having it in the .env.example it will communicate the requirement for a name claim and also let you set it to something like preferred_username when the name claim is not available or configurable.

**Comment by ghost** [comment]
Also, I hate to bring this up, but `passport-oauth2` is not actually suitable for implementing OIDC.

1. OIDC's main feature is ID tokens, but `passport-oauth2` only returns the `typ: Bearer` token (aka "access_token").
2. OIDC's other feature is discovery of endpoints (and being able to use a registration token to bootstrap a client), one should only need to specify the issuer. The discovery endpoint can be derived by adding `.well-known/openid-configuration`, which contains all the endpoints that have to be  specified manually right now (auth/token/userinfo)
3. The UserInfo endpoint is not for pulling a specific set of claims, but is usually used when the claims would be too large for the token (e.g. we had a case of 200+ permission flags being in a claim). Preferably the ID token should be used.
4. The Scope `openid` is required for a request to include an ID token, and can also effect UserInfo output. This is the "toggle" in OAuth2 to turn on OIDC and should always be included, otherwise this implementation should be called OAuth2.

I'd be happy to contribute an OIDC implementation (also possibly with groups claims), but I don't think it can be done fully backwards compatible. But then again, the current requirements (having your entire profile in userinfo) aren't well documented either. The OIDC spec unfortunately did not give strong guarantees about any but the most basic claims (not even `azp`).

**Reply by tommoor** [reply]
> OIDC's other feature is discovery of endpoints

It would be great to have this included. I think we care less about the semantics of being strictly OIDC and more about having a solution that works for as many usecases as possible with the least amount of configuration. I'm no OIDC expert so this part of the project is very much community driven
